### PR TITLE
Use native isoformat/fromisoformat

### PR DESCRIPTION
**Pull request by Pliner**
The PR replaces iso parsing/formatting function with built in one, which were fixed starting from python 3.11.

27% time decrease for marshmallow 2.21, 17% time decrease for mashmallow 3.21.

### CODE REVIEW

**Comment by Pliner** [comment]
python 3.12, marshmallow 3.21

before:
```
./benchmarks/transaction_load_dump.py: 
         307263 function calls (304191 primitive calls) in 0.097 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   2050/2    0.011    0.000    0.075    0.038 schema.py:575(_deserialize)
     2048    0.007    0.000    0.014    0.000 utils.py:145(from_iso_datetime)
    10240    0.006    0.000    0.049    0.000 fields.py:342(deserialize)
     6144    0.004    0.000    0.006    0.000 fields.py:1091(_format_num)
     8192    0.003    0.000    0.009    0.000 fields.py:263(_validate)
     8192    0.003    0.000    0.004    0.000 fields.py:269(_validate_all)
    10240    0.003    0.000    0.052    0.000 schema.py:654(getter)
     2048    0.003    0.000    0.005    0.000 uuid.py:139(__init__)
     5120    0.002    0.000    0.016    0.000 fields.py:315(serialize)
     2048    0.002    0.000    0.003    0.000 utils.py:135(get_fixed_timezone)
     4096    0.002    0.000    0.008    0.000 fields.py:1101(_validated)
    22530    0.002    0.000    0.003    0.000 {built-in method builtins.isinstance}
   1025/1    0.002    0.000    0.018    0.018 schema.py:500(_serialize)
    10240    0.002    0.000    0.054    0.000 schema.py:480(_call_and_store)
    13318    0.002    0.000    0.002    0.000 {method 'get' of 'dict' objects}
     6144    0.002    0.000    0.002    0.000 {method 'quantize' of 'decimal.Decimal' objects}
     2048    0.002    0.000    0.002    0.000 {method 'match' of 're.Pattern' objects}
     5120    0.001    0.000    0.004    0.000 utils.py:262(get_value)
     2048    0.001    0.000    0.003    0.000 inspect.py:384(_has_code_flag)
     2048    0.001    0.000    0.001    0.000 {method 'astimezone' of 'datetime.datetime' objects}
     1024    0.001    0.000    0.001    0.000 {method 'isoformat' of 'datetime.datetime' objects}
     2048    0.001    0.000    0.017    0.000 fields.py:660(_deserialize)
        8    0.001    0.000    0.003    0.000 schema.py:1195(_invoke_processors)
     8192    0.001    0.000    0.001    0.000 validate.py:66(__init__)
     2048    0.001    0.000    0.006    0.000 fields.py:910(_validated)
    10240    0.001    0.000    0.001    0.000 utils.py:298(set_value)
     5120    0.001    0.000    0.002    0.000 utils.py:288(_get_value_for_key)
     5120    0.001    0.000    0.005    0.000 fields.py:251(get_value)
     2048    0.001    0.000    0.001    0.000 {method 'groupdict' of 're.Match' objects}
     6144    0.001    0.000    0.001    0.000 {method 'replace' of 'str' objects}
     2048    0.001    0.000    0.015    0.000 fields.py:1281(_deserialize)
     4096    0.001    0.000    0.005    0.000 fields.py:950(_validated)
     8192    0.001    0.000    0.001    0.000 validate.py:73(__call__)
     4096    0.001    0.000    0.009    0.000 fields.py:974(_deserialize)
     5120    0.001    0.000    0.004    0.000 schema.py:468(get_attribute)
    10240    0.001    0.000    0.001    0.000 fields.py:306(_validate_missing)
     2048    0.001    0.000    0.005    0.000 utils.py:46(is_generator)
     1024    0.001    0.000    0.001    0.000 {method 'replace' of 'datetime.datetime' objects}
     2048    0.001    0.000    0.004    0.000 fields.py:967(_serialize)
     6144    0.001    0.000    0.001    0.000 {method 'items' of 'dict' objects}
     1024    0.001    0.000    0.001    0.000 <string>:2(__init__)
     6144    0.001    0.000    0.001    0.000 {method 'is_finite' of 'decimal.Decimal' objects}
     2050    0.001    0.000    0.006    0.000 utils.py:51(is_iterable_but_not_string)
     2048    0.001    0.000    0.003    0.000 fields.py:666(_serialize)
     7172    0.001    0.000    0.001    0.000 {built-in method builtins.hasattr}
     1024    0.001    0.000    0.001    0.000 uuid.py:280(__str__)
     2048    0.000    0.000    0.000    0.000 {built-in method builtins.format}
     2048    0.000    0.000    0.006    0.000 fields.py:923(_deserialize)
     2049    0.000    0.000    0.000    0.000 typing.py:371(inner)
     4096    0.000    0.000    0.000    0.000 {method 'is_nan' of 'decimal.Decimal' objects}
     2050    0.000    0.000    0.006    0.000 utils.py:56(is_collection)
     5124    0.000    0.000    0.000    0.000 {built-in method builtins.getattr}
     2048    0.000    0.000    0.004    0.000 inspect.py:395(isgeneratorfunction)
     4096    0.000    0.000    0.000    0.000 {method 'is_infinite' of 'decimal.Decimal' objects}
     1024    0.000    0.000    0.002    0.000 fields.py:1272(_serialize)
     1024    0.000    0.000    0.001    0.000 utils.py:242(ensure_text_type)
     4096    0.000    0.000    0.000    0.000 {built-in method builtins.len}
     2050    0.000    0.000    0.001    0.000 <frozen abc>:117(__instancecheck__)
     1024    0.000    0.000    0.001    0.000 bake.py:290(post_load)
     1024    0.000    0.000    0.000    0.000 bake.py:284(remove_none_values)
     2048    0.000    0.000    0.000    0.000 functools.py:421(_unwrap_partial)
     2048    0.000    0.000    0.000    0.000 inspect.py:2135(_signature_is_functionlike)
     4096    0.000    0.000    0.000    0.000 {built-in method builtins.callable}
     2048    0.000    0.000    0.000    0.000 inspect.py:299(ismethod)
     2050    0.000    0.000    0.000    0.000 {built-in method _abc._abc_instancecheck}
     2048    0.000    0.000    0.000    0.000 inspect.py:371(isfunction)
     2048    0.000    0.000    0.000    0.000 bake.py:294(pre_load)
     2048    0.000    0.000    0.000    0.000 inspect.py:440(isgenerator)
     2048    0.000    0.000    0.001    0.000 fields.py:1111(_to_string)
     4097    0.000    0.000    0.000    0.000 typing.py:2107(cast)
     2048    0.000    0.000    0.000    0.000 {method 'pop' of 'dict' objects}
     2048    0.000    0.000    0.000    0.000 {method 'count' of 'list' objects}
     2048    0.000    0.000    0.000    0.000 {method 'strip' of 'str' objects}
     2048    0.000    0.000    0.000    0.000 {method 'ljust' of 'str' objects}
     3072    0.000    0.000    0.000    0.000 schema.py:413(dict_class)
     2048    0.000    0.000    0.000    0.000 {built-in method builtins.divmod}
     2048    0.000    0.000    0.000    0.000 {built-in method builtins.abs}
     1024    0.000    0.000    0.001    0.000 fields.py:890(_serialize)
     1024    0.000    0.000    0.001    0.000 utils.py:226(isoformat)
        1    0.000    0.000    0.097    0.097 <string>:1(<module>)
        1    0.000    0.000    0.039    0.039 schema.py:778(validate)
        1    0.000    0.000    0.097    0.097 {built-in method builtins.exec}
        2    0.000    0.000    0.078    0.039 schema.py:810(_do_load)
        1    0.000    0.000    0.019    0.019 schema.py:524(dump)
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
        1    0.000    0.000    0.058    0.058 serialization.py:91(dump_many_v3)
        2    0.000    0.000    0.000    0.000 serialization.py:57(schema_v3)
        1    0.000    0.000    0.001    0.001 schema.py:1062(_invoke_dump_processors)
        3    0.000    0.000    0.002    0.001 schema.py:1076(_invoke_load_processors)
        1    0.000    0.000    0.039    0.039 schema.py:691(load)
        1    0.000    0.000    0.039    0.039 serialization.py:73(load_many_v3)
        2    0.000    0.000    0.000    0.000 <string>:2(__eq__)
        2    0.000    0.000    0.000    0.000 schema.py:1105(_invoke_field_validators)
        2    0.000    0.000    0.000    0.000 <string>:2(__hash__)
        2    0.000    0.000    0.000    0.000 {built-in method builtins.hash}
        2    0.000    0.000    0.000    0.000 error_store.py:13(__init__)
```

after:

```
./benchmarks/transaction_load_dump.py:22: 
   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
   2050/2    0.011    0.000    0.060    0.030 schema.py:575(_deserialize)
    10240    0.006    0.000    0.033    0.000 fields.py:342(deserialize)
     6144    0.004    0.000    0.006    0.000 fields.py:1091(_format_num)
     8192    0.003    0.000    0.009    0.000 fields.py:263(_validate)
     8192    0.003    0.000    0.004    0.000 fields.py:269(_validate_all)
    10240    0.003    0.000    0.036    0.000 schema.py:654(getter)
     2048    0.003    0.000    0.004    0.000 uuid.py:139(__init__)
     5120    0.003    0.000    0.016    0.000 fields.py:315(serialize)
   1025/1    0.002    0.000    0.019    0.019 schema.py:500(_serialize)
     4096    0.002    0.000    0.008    0.000 fields.py:1101(_validated)
    20482    0.002    0.000    0.003    0.000 {built-in method builtins.isinstance}
    10240    0.002    0.000    0.038    0.000 schema.py:480(_call_and_store)
    13318    0.002    0.000    0.002    0.000 {method 'get' of 'dict' objects}
     6144    0.002    0.000    0.002    0.000 {method 'quantize' of 'decimal.Decimal' objects}
     1024    0.001    0.000    0.001    0.000 {method 'isoformat' of 'datetime.datetime' objects}
     5120    0.001    0.000    0.004    0.000 utils.py:262(get_value)
     2048    0.001    0.000    0.003    0.000 inspect.py:384(_has_code_flag)
     2048    0.001    0.000    0.003    0.000 fields.py:660(_deserialize)
     8192    0.001    0.000    0.001    0.000 validate.py:66(__init__)
        8    0.001    0.000    0.003    0.000 schema.py:1195(_invoke_processors)
    10240    0.001    0.000    0.001    0.000 utils.py:298(set_value)
     2048    0.001    0.000    0.006    0.000 fields.py:910(_validated)
     5120    0.001    0.000    0.005    0.000 fields.py:251(get_value)
     5120    0.001    0.000    0.002    0.000 utils.py:288(_get_value_for_key)
     6144    0.001    0.000    0.001    0.000 {method 'replace' of 'str' objects}
     8192    0.001    0.000    0.001    0.000 validate.py:73(__call__)
     4096    0.001    0.000    0.005    0.000 fields.py:950(_validated)
     4096    0.001    0.000    0.009    0.000 fields.py:974(_deserialize)
     5120    0.001    0.000    0.004    0.000 schema.py:468(get_attribute)
    10240    0.001    0.000    0.001    0.000 fields.py:306(_validate_missing)
     2048    0.001    0.000    0.001    0.000 fields.py:1281(_deserialize)
     1024    0.001    0.000    0.001    0.000 {method 'replace' of 'datetime.datetime' objects}
     2048    0.001    0.000    0.004    0.000 fields.py:967(_serialize)
     1024    0.001    0.000    0.001    0.000 <string>:2(__init__)
     2048    0.001    0.000    0.005    0.000 utils.py:46(is_generator)
     6144    0.001    0.000    0.001    0.000 {method 'is_finite' of 'decimal.Decimal' objects}
     2048    0.001    0.000    0.003    0.000 fields.py:666(_serialize)
     1024    0.001    0.000    0.001    0.000 uuid.py:280(__str__)
     2050    0.001    0.000    0.005    0.000 utils.py:51(is_iterable_but_not_string)
     7172    0.001    0.000    0.001    0.000 {built-in method builtins.hasattr}
     4096    0.000    0.000    0.000    0.000 {method 'items' of 'dict' objects}
     5124    0.000    0.000    0.000    0.000 {built-in method builtins.getattr}
     2048    0.000    0.000    0.000    0.000 {built-in method builtins.format}
     2048    0.000    0.000    0.006    0.000 fields.py:923(_deserialize)
     4096    0.000    0.000    0.000    0.000 {method 'is_nan' of 'decimal.Decimal' objects}
     2049    0.000    0.000    0.000    0.000 typing.py:371(inner)
     2050    0.000    0.000    0.006    0.000 utils.py:56(is_collection)
     2048    0.000    0.000    0.003    0.000 inspect.py:395(isgeneratorfunction)
     4096    0.000    0.000    0.000    0.000 {method 'is_infinite' of 'decimal.Decimal' objects}
     2048    0.000    0.000    0.000    0.000 {built-in method fromisoformat}
     1024    0.000    0.000    0.001    0.000 utils.py:242(ensure_text_type)
     1024    0.000    0.000    0.002    0.000 fields.py:1272(_serialize)
     1024    0.000    0.000    0.000    0.000 bake.py:284(remove_none_values)
     1024    0.000    0.000    0.001    0.000 bake.py:290(post_load)
     2050    0.000    0.000    0.001    0.000 <frozen abc>:117(__instancecheck__)
     2048    0.000    0.000    0.000    0.000 functools.py:421(_unwrap_partial)
     2048    0.000    0.000    0.000    0.000 {method 'astimezone' of 'datetime.datetime' objects}
     2048    0.000    0.000    0.000    0.000 inspect.py:2135(_signature_is_functionlike)
     4096    0.000    0.000    0.000    0.000 {built-in method builtins.callable}
     2050    0.000    0.000    0.000    0.000 {built-in method _abc._abc_instancecheck}
     2048    0.000    0.000    0.000    0.000 inspect.py:299(ismethod)
     2048    0.000    0.000    0.000    0.000 inspect.py:440(isgenerator)
     2048    0.000    0.000    0.000    0.000 inspect.py:371(isfunction)
     2048    0.000    0.000    0.001    0.000 fields.py:1111(_to_string)
     2048    0.000    0.000    0.000    0.000 bake.py:294(pre_load)
     4097    0.000    0.000    0.000    0.000 typing.py:2107(cast)
     2048    0.000    0.000    0.000    0.000 {method 'count' of 'list' objects}
     2048    0.000    0.000    0.000    0.000 {method 'strip' of 'str' objects}
     3072    0.000    0.000    0.000    0.000 schema.py:413(dict_class)
     1024    0.000    0.000    0.001    0.000 fields.py:890(_serialize)
     2048    0.000    0.000    0.000    0.000 {built-in method builtins.len}
        1    0.000    0.000    0.081    0.081 <string>:1(<module>)
        1    0.000    0.000    0.031    0.031 schema.py:778(validate)
        1    0.000    0.000    0.081    0.081 {built-in method builtins.exec}
        1    0.000    0.000    0.019    0.019 schema.py:524(dump)
        2    0.000    0.000    0.062    0.031 schema.py:810(_do_load)
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
        1    0.000    0.000    0.050    0.050 serialization.py:91(dump_many_v3)
        2    0.000    0.000    0.000    0.000 serialization.py:57(schema_v3)
        1    0.000    0.000    0.001    0.001 schema.py:1062(_invoke_dump_processors)
        3    0.000    0.000    0.002    0.001 schema.py:1076(_invoke_load_processors)
        1    0.000    0.000    0.031    0.031 schema.py:691(load)
        2    0.000    0.000    0.000    0.000 <string>:2(__eq__)
        1    0.000    0.000    0.031    0.031 serialization.py:73(load_many_v3)
        2    0.000    0.000    0.000    0.000 <string>:2(__hash__)
        2    0.000    0.000    0.000    0.000 schema.py:1105(_invoke_field_validators)
        2    0.000    0.000    0.000    0.000 {built-in method builtins.hash}
        2    0.000    0.000    0.000    0.000 error_store.py:13(__init__)
```

File: marshmallow_recipe/fields.py [hunk 1]
```diff
@@ -633,10 +634,35 @@ def _deserialize(self, value: Any, attr: Any, data: Any, **kwargs: Any) -> Any:
...
+                if sys.version_info >= (3, 11)
+                else {}
+            ),
+        }
+
         def _deserialize(self, value: Any, attr: Any, data: Any, **kwargs: Any) -> Any:
             result = super()._deserialize(value, attr, data, **kwargs)
             if result.tzinfo is None:
                 return result.replace(tzinfo=datetime.timezone.utc)
+            if result.tzinfo == datetime.timezone.utc:
```
Comment (Pliner) [inline marshmallow_recipe/fields.py:664]: No need to call astimezone(utc) if tzinfo is utc

File: marshmallow_recipe/fields.py [hunk 2]
```diff
@@ -834,10 +860,50 @@ def _deserialize(self, value: Any, attr: Any, data: Any, **_: Any) -> Any:
...
+            if value.tzinfo is None:
+                value = value.replace(tzinfo=datetime.timezone.utc)
+
+            return super()._serialize(value, attr, obj, **kwargs)
+
         def _deserialize(self, value: Any, attr: Any, data: Any, **_: Any) -> Any:
             result = super()._deserialize(value, attr, data)
             if result.tzinfo is None:
                 return result.replace(tzinfo=datetime.timezone.utc)
+            if result.tzinfo == datetime.timezone.utc:
```
Comment (Pliner) [inline marshmallow_recipe/fields.py:905]: No need to call astimezone(utc) if tzinfo is utc

File: marshmallow_recipe/fields.py [hunk 3]
```diff
@@ -834,10 +860,50 @@ def _deserialize(self, value: Any, attr: Any, data: Any, **_: Any) -> Any:
...
+                if sys.version_info >= (3, 11)
+                else {}
+            ),
+        }
+
+        def _serialize(self, value: Any, attr: Any, obj: Any, **kwargs: Any) -> Any:
+            if value is None:
+                return None
+
+            if value.tzinfo is None:
```
Comment (Pliner) [inline marshmallow_recipe/fields.py:896]: missing part for marshmallow 2: date-utils was used and it ultimately auto converted datetime to utc if no timezone is specified.

**Review by fedorfo — APPROVED** [review]

**Review by fedorfo — APPROVED** [review]

**Pull request merged**